Exploration RECC complaints data

sofiapinto opened the pull request:
Closes #7 

Hey @ch-williamson,

Thank you so much for taking the time to review my code. This code processes RECC complaints data and does the following analysis: descriptive analysis, specified keyword and expression analysis and top n-gram analysis.

Could you please make sure that the following scripts run without error and that the figures are storing to the correct folders?
- `python3 /exploration_recc_complaints/processing_recc_data.py`
- `python3 /exploration_recc_complaints/descriptive_analysis.py`
- `python3 /exploration_recc_complaints/keyword_analysis.py`
- `python3 /exploration_recc_complaints/top_ngrams_analysis.py`

I would also appreciate any comments/suggestions you might have on the scripts above, as well as if anything important might be missing from this exploration's `README.md` file.

I have also created a notebook (although saved as a .py file) which allows you to look at the raw data and processed data - which might be helpful while reviewing.

Because this is an explorations repo, we're not using the cookiecutter structure. However, I can create some kind of structure under `/exploration_recc_complaints/`, with a `pipeline`, `analysis`, `utils` and `getters` folders. Would you find that helpful? 

**Setup for review:**
1. Clone this repo: `git clone git@github.com:nestauk/asf_exploration.git`
2. Checkout the correct branch: `git checkout 7_recc_complaints`
3. Navigate to this exploration's folder: `cd asf_exploration/exploration_recc_complaints`
4. Create your conda environment: `conda create --name exploration_recc_complaints python=3.9`
5. Activate your conda environment: `conda activate exploration_recc_complaints`
6. Install package dependencies: `pip3 install -r requirements.txt`
7. Add your conda environment to the notebooks: `python3 -m ipykernel install --user --name=exploration_recc_complaints`

Thanks a lot in advance!
Sofia

Commit:
sub directory with readme file

sofiapinto committed:
adding script to process data, getters and config

sofiapinto committed:
change to procesing script

sofiapinto committed:
adding descriptive analysis

sofiapinto committed:
changes to data processing and descriptive analysis

sofiapinto committed:
adding text analysis scripts

sofiapinto committed:
replace mkdir by makedirs and black formatting

sofiapinto committed:
add project name and description to main README.md

sofiapinto committed:
removing unecessary line

sofiapinto committed:
updating paths

sofiapinto committed:
solving paths bug

sofiapinto committed:
formatting prints

sofiapinto committed:
tidying code

sofiapinto committed:
removing file

sofiapinto committed:
Update README.md

ch-williamson commented on exploration_recc_complaints/processing_recc_data.py:
```diff
@@ -0,0 +1,207 @@
+categories_short_names = config.categories_short_names
+
+
+def camel_case_columns(df: pd.DataFrame):
```
I think this is actually snake case! Camel is likeThis

sofiapinto replied:
upsi - thanks Chris!

ch-williamson commented on exploration_recc_complaints/general_utils.py:
```diff
@@ -0,0 +1,86 @@
+    )
+
+
+def lemmatising(tokens: list[str]) -> str:
```
I'm not an NLP expert, but I had a look at the lemmatised comments and was surprised to find "has" being lemmatised to "ha". I would assume that it would interpret "has" as a verb by default but apparently not...

`lemmatizer.lemmatize("has", pos="v")` returns "have" as expected. I guess you could add it as a special case here but I feel like there must be a way of getting NLTK to use the most likely POS tag by default - any ideas?

This specific example is a stopword, so it's removed anyway, but just wondering if it's indicative of a wider problem!

sofiapinto replied:
Thanks Chris, I noticed that too, and it's definitely something we should look into. First thing would be to compare to outputs from other packages, to check if it's a problem of NLTK. For now, because this project is not going forward, I think I'll add it to the README.md as a problem not currently addressed and also to the summary doc - so that we remember to address it if we pick this up later.

ch-williamson commented on exploration_recc_complaints/processing_recc_data.py:
```diff
@@ -0,0 +1,207 @@
+
+    """
+
+    unique_values = list(set(data[variable].unique().sum().split("; ")[:-1]))
```
I get what this is doing - but a quick comment might help explain for future reference!

sofiapinto replied:
added, thanks :)

ch-williamson commented on exploration_recc_complaints/processing_recc_data.py:
```diff
@@ -0,0 +1,207 @@
+
+    dummy_variable_names = [prefix + ":" + c for c in unique_values]
+
+    for c in range(len(unique_values)):
```
Consider using a different variable here as `c` refers to a string above

sofiapinto replied:
will do, to avoid confusion :) although `c` doesn't exist outside `dummy_variable_names`, so wouldn't be a problem to use it again (but I see that we are using it do define two different things)

ch-williamson commented on exploration_recc_complaints/processing_recc_data.py:
```diff
@@ -0,0 +1,207 @@
+    Returns:
+        Processed dataframe.
+    """
+    data["categories"].fillna("No category specified; ", inplace=True)
```
In the case of the tech columns, is there a risk that this will lead to `number_of_technologies` being 1 even if there were no technologies involved (i.e. the value was empty)?

sofiapinto replied:
Not for tech columns, as this function is only applied to categories. But it will have `number_of_categories` as 1 if "No category specified; ", so I'll fix that so that it is 0. Thanks for spotting it @ch-williamson  👍

sofiapinto replied:
Also added plots to look at these distributions under the descriptive analysis script:

- percentage of complaints with a given number of categories (`percentage_complaints_by_number_of_categories.png`)
- percentage of complaints with a given number of technologies (`percentage_complaints_by_number_of_technologies.png`)

ch-williamson commented on exploration_recc_complaints/visualisation_utils.py:
```diff
@@ -0,0 +1,222 @@
+    dfi.export(df, folder + figure_name)
+
+
+def max_value_to_show(values: pd.Series) -> list[int]:
```
Handy function, cleverly implemented :)

sofiapinto replied:
thanks @ch-williamson! :)

ch-williamson commented on exploration_recc_complaints/descriptive_analysis.py:
```diff
@@ -0,0 +1,237 @@
+            density=True,
+        )
+    ax[i].set_xlabel("Complaint length")
+    fig.suptitle("Density of complaints yearly")
```
Could perhaps be made a bit more clear - maybe title "Density/Distribution of complaint lengths by year", x-axis "Number of characters"?

ch-williamson commented on exploration_recc_complaints/visualisation_utils.py:
```diff
@@ -0,0 +1,222 @@
+import seaborn as sns
+from matplotlib import font_manager
+import pandas as pd
+import dataframe_image as dfi
```
Have never come across this before but it's cool!

sofiapinto replied:
me neither, before this project :)

ch-williamson commented on exploration_recc_complaints/keyword_analysis.py:
```diff
@@ -0,0 +1,135 @@
+import pandas as pd
+import os
+import visualisation_utils
+import matplotlib.pyplot as plt
```
Looks like you don't use this here

sofiapinto replied:
thanks! removed :)

ch-williamson commented on exploration_recc_complaints/keyword_analysis.py:
```diff
@@ -0,0 +1,135 @@
+        columns = columns + [
+            "EXP:" + col for col in keywords_expressions[g] if " " in col
+        ]
+        data["G:" + g] = data[columns].sum(axis=1)
+
+    group_columns = [col for col in data.columns if col.startswith("G:")]
+    for col in group_columns:
+        data[col] = data[col].apply(lambda x: 1 if x > 1 else x)
```
Could this just be `data["G:" + g] = data[columns].any(axis=1).astype(int)`?

sofiapinto replied:
yessss! :) thank you

ch-williamson commented on exploration_recc_complaints/keyword_analysis.py:
```diff
@@ -0,0 +1,135 @@
+        data[col] = data[col].apply(lambda x: 1 if x > 1 else x)
+
+
+def performs_keyword_expression_analysis(data):
```
Picky thing, but I think recommended style is to use second person imperative verbs in function names (i.e. this would be "perform_..."

sofiapinto replied:
thanks :) changed it!

ch-williamson commented on exploration_recc_complaints/visualisation_utils.py:
```diff
@@ -0,0 +1,222 @@
+
+    set_spines()
+
+    # Had trouble making it find the font I set so this was the only way to do it
+    # without specifying the local filepath
+    font_files = font_manager.findSystemFonts(fontpaths=["/Users/"])
+
+    for font_file in font_files:
+        font_manager.fontManager.addfont(font_file)
```
I got an error here about not being able to set the fontsize for "NotoColorEmoji.ttf". You could do:
```
for font_file in font_files:
	if (config.font in font_file) or (config.title_font in font_file):
		font_manager.fontManager.addfont(font_file)
```
(after importing config). You could even wrap this in a try/except that prompts the user to check that they have the relevant fonts installed.

sofiapinto replied:
I went with a slightly different solution, but hopefully it will work for you now :)

ch-williamson reviewed (approved):
Hi @sofiapinto - I've reviewed and all looks great :) have only left a few minor comments.

I ran the scripts you listed - `processing_recc_data.py` worked fine, but the rest failed with this error:
```
  File "/Users/chris.williamson/asf_exploration/exploration_recc_complaints/visualisation_utils.py", line 179, in <module>
    data: str | pd.DataFrame,
TypeError: unsupported operand type(s) for |: 'type' and 'type'
```

This is because the | in type hints only works in Python 3.10+ and I was on 3.9. After upgrading I got this error:
```
  File "/Users/chris.williamson/asf_exploration/exploration_recc_complaints/visualisation_utils.py", line 43, in set_plotting_styles
    font_manager.fontManager.addfont(font_file)
  File "/Users/chris.williamson/opt/anaconda3/envs/exploration_recc_complaints/lib/python3.10/site-packages/matplotlib/font_manager.py", line 1132, in addfont
    font = ft2font.FT2Font(path)
RuntimeError: In FT2Font: Could not set the fontsize (invalid pixel size; error code 0x17)
```

Turns out I have a couple of problematic fonts that break things - I've made a suggestion for fixing this. After this, all the scripts worked fine although `descriptive_analysis.py` threw a few warnings that looked like
```
objc[4798]: Class WebSwapCGLLayer is implemented in both /System/Library/Frameworks/WebKit.framework/Versions/A/Frameworks/WebCore.framework/Versions/A/Frameworks/libANGLE-shared.dylib (0x7ffb43640ec8) and /Applications/Google Chrome.app/Contents/Frameworks/Google Chrome Framework.framework/Versions/109.0.5414.87/Libraries/libGLESv2.dylib (0x113e80880). One of the two will be used. Which one is undefined.
```

Not sure what this is about, but it didn't seem to break anything - all the outputs looked fine and were in the right places.

sofiapinto committed:
addressing PR suggestions

sofiapinto committed:
plots for distribution of number of cats and techs

sofiapinto committed:
correcting small typos

sofiapinto committed:
adding english stopwords function

sofiapinto committed:
updating README.md

sofiapinto committed:
replace .py by .ipynb

sofiapinto committed:
remove debugging print

sofiapinto committed:
changing plot title and fixing typo

sofiapinto commented:
Hey @ch-williamson, thank you so much for reviewing the code and for the useful suggestions!

- I have addressed your suggestions;
- Regarding this warning `objc[4798]: Class WebSwapCGLLayer is implemented in both /System/Library/Frameworks/WebKit.framework/Versions/A/Frameworks/WebCore.framework/Versions/A/Frameworks/libANGLE-shared.dylib (0x7ffb43640ec8) and /Applications/Google Chrome.app/Contents/Frameworks/Google Chrome Framework.framework/Versions/109.0.5414.87/Libraries/libGLESv2.dylib (0x113e80880). One of the two will be used. Which one is undefined.
`, I also get something similar. It comes from using the `dataframe_image` package, but all seems to work well, so I'll just ignore it.
- I have also added two plots to the descriptive analysis that I forgot when refactoring, showing the percentage of complaints with a certain number of categories/technologies. 
- Fixed extra typos I found;
- Improved the `README.md` by adding extra info and writing down the current problem with NLTK's lemmatiser.

Can you please re-run the scripts to check if all still runs correctly for you after the changes? Hopefully, you no longer have problems with fonts now that I changed the code.

Thanks so much.

sofiapinto committed:
fixing problems with fonts
